rep-2004.rst

```diff
@@ -0,0 +1,527 @@
+  * packages which are needed for production systems
+  * e.g. ``rclcpp``, ``urdf``, ``tf2``, etc.
+
+* Quality Level 2:
+
+  * high quality packages which are either:
+
+    * on the way to level 1 or
+    * are general solutions used by many people, but are only sometimes used for production systems
+
+  * e.g. ``navigation2``, ``rosbag2``, etc.
+
+* Quality Level 3:
+
+  * tooling quality packages
+  * e.g. ``ros2cli``, ``rviz``, ``rqt``, etc.
+
+* Quality Level 4:
+
+  * demos, tutorials, and experiments
+  * e.g. research packages, ``demo_nodes_cpp``, ``examples_rclcpp_minimal_publisher``, etc.
+
+While each quality level will have different requirements, it's always possible to overachieve in certain requirements even if other requirements prevent a package from moving up to the next quality level.
+
+Quality Level 1
+^^^^^^^^^^^^^^^
+
+This category should be used for packages which are required for a reasonable ROS system in a production environment.
+That is to say that after you remove development tools, build tools, and introspection tools, these packages are still left over as requirements for a basic ROS system to run.
+However, that does not mean that packages that would not normally fit this description should never be called 'Level 1'.
+If there is a need for a particular package in a reasonable production scenario, then that package should be considered for this category as well.
+However, packages which we consider essential to getting a robot up and running quickly, but perhaps is a generic solution to the problem should probably not start out as 'Level 1' due to the high effort in getting a package to 'Level 1' and maintaining it there.
+
+For example, the packages which provide intra-process communication, inter-process communication, generated message runtime code, node lifecycle, etc. should probably all be considered for 'Level 1'.
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+However, it may upgrade to it at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+*Note: bullets below that start with [ROS Core], will be the prescription for what we do in the core packages in order to meet the associated requirements*
+
+Requirements to be considered a 'Level 1' package:
+
+* Version Policy:
+
+  * Must have a version policy (e.g. ``semver``)
+  * Must be at a stable version (e.g. for ``semver`` that means have a version >= 1.0.0)
```

Comment:
I had that originally, but I wanted this to be more inclusive. Some existing projects already have versioning schemes and won't be convinced to change. I don't think we need any requirements for the version policy, just that it exists and is well understood by the consumers (i.e. documented).

The added inclusiveness is helpful if you want to try and apply these requirements to an external dependency (e.g. something like opencv or eigen).

I'd leave this as-is unless you or someone else feels strongly about it.

Reply:
Does this REP target ROS packages, or "all" software in "the ROS ecosystem"? If the former: I would say we could require what makes sense. For non-ROS packages, it would not seem necessary.

The REP later on already allows "explaining" why a non-ROS dependency qualifies for a certain level even if it doesn't follow all requirements put forth here. It doesn't seem necessary for this section to be inclusive of other versioning schemes then.

Reply:
tl;dr Do we really want semver for the core?

---

ROS core packages haven't historically _really_ used semver. And that's probably ok and by design, so maybe we should think twice about trying to introduce semver into ROS 2 core. It seems that the way of communicating stability in the core is to be within a distribution, regardless of whether the major version is still 0.

If ROS maintainers don't use semver to communicate compatibility, maybe we don't want or need semver. Let's take `rclcpp` for example. If we tell a downstream consumer that it is following semver, they look at version 0 and assume no guarantees are made about stability, but that's not true. Version `0.7.7` is guaranteed to be API compatible with `0.7.11`, because they're both on Dashing. The same goes for `0.8.0` and `0.8.1`, both on Eloquent. It seems that minors are being treated as semver's majors.

Talking to @dirk-thomas, it seems like version 1.0.0 is not waiting for API stability, but for a deep API review. It almost feels like on the "**rosver™**"  scheme, a version 2.0.0 of a core package would warrant a refactor into ROS 3 :sweat_smile: 

If we're not using majors to communicate API compatibility, minors to communicate backwards compatible feature additions and patches to communicate bug fixes, why use semver?

Reply:
I'd say we're following semver, albeit conservatively at the moment. Currently minor versions are coupled with the distro and once we bump to 1.0.0, the major version will be coupled with the distro.

Reply:
I think discussions have happened elsewhere in the meantime, but just to tie this end up...

---

> > The added inclusiveness is helpful if you want to try and apply these requirements to an external dependency (e.g. something like opencv or eigen).

> But those are not ROS packages?

No, but that was just one example. I think the more compelling example (now that I think about it more) is that of something like `cartographer_ros`, which is a ROS package, but comes from another community which maybe prefers a different versioning scheme (that may not be the case, but just as an example). Who's to say they should use semver in their wrapper and not the versioning scheme of the thing that they wrap? I think it would be reasonable for them to do so, but acceptable if they choose not to, so long as their version scheme is sufficient. If anything, structuring it like it is just clarifies what it is about semver that is valuable in this document. This serves to underscore why a version scheme like semver is useful here and it also makes it easier for developers trying to justify a different version scheme if they feel they need to do so.

I definitely see what your perspective, and I think it is valid, but personally I prefer the flexibility of how it is worded now, and for all intents and purposes I believe users will take this away as "use semver", as that will the recommendation in our developer guide, and that is what we will use (many people just copy the standards of core packages when setting their own up). So again, leaving it as-is would be my preference. But if there are others that agree with you instead please speak up and I'll reconsider it.

---

> tl;dr Do we really want semver for the core?

I think so.

`semver` has more to do with "if the version changed, what can I imply from the delta?" and less to do with "what can be changed in a single ROS distribution". Also, `semver` is another "familiar" thing for new developers and one less "special ROS thing" they have to put up with. So, I think it's a good thing to do (but not absolutely required) to move to 1.0.0 and follow semver.

Whether or not we use semver, we'll still have conventions about what changes within a ROS distribution as we do now.